tests/test_asvdb.py

```diff
@@ -279,6 +279,110 @@ def test_concurrency_stress():
+
+    bInfo = BenchmarkInfo()
+    bResult1 = BenchmarkResult(funcName="somebenchmark1", result=43)
+    bResult2 = BenchmarkResult(funcName="somebenchmark2", result=43)
+    bResult3 = BenchmarkResult(funcName="somebenchmark3", result=43)
+
+    # db1 or db2 should be actively writing the result (because the writeDelay is long)
+    # and db3 should be blocked.
+    t1 = threading.Thread(target=db1.addResult, args=(bInfo, bResult1))
+    t2 = threading.Thread(target=db2.addResult, args=(bInfo, bResult2))
+    t3 = threading.Thread(target=db3.addResult, args=(bInfo, bResult3))
+    t1.start()
+    t2.start()
+    time.sleep(0.5)  # ensure t3 tries to write last
+    t3.start()
+
+    # Check that db3 is blocked - if locking wasn't working, it would have
+    # finished since it has no writeDelay.
+    t3.join(timeout=0.5)
+    assert t3.is_alive() is True
+
+    # Cancel db1 and db2, allowing db3 to write and finish
+    db1.cancelWrite = True
+    db2.cancelWrite = True
+    t3.join(timeout=11)
+    assert t3.is_alive() is False
+    t1.join()
+    t2.join()
+    t3.join()
+
+    # Check that db3 wrote its result
+    os.makedirs(path.join(tmpDir.name, "asvdb/results"))
+    resource.Bucket(bucketName).download_file(benchmarkKey, path.join(tmpDir.name, benchmarkKey))
+    with open(path.join(tmpDir.name, benchmarkKey)) as fobj:
+        jo = json.load(fobj)
+        assert "somebenchmark3" in jo
+
+    tmpDir.cleanup()
+    db3.s3Resource.Bucket(db3.bucketName).objects.filter(Prefix="asvdb/").delete()
+
+
+def test_s3_concurrency_stress():
+    from asvdb import ASVDb, BenchmarkInfo, BenchmarkResult
+
+    asvDirName = "s3://gpuci-cache-testing/asvdb"
+    bucketName = "gpuci-cache-testing"
+    repo = "somerepo"
+    branch1 = "branch1"
+    num = 32
+    dbs = []
+    threads = []
+    allFuncNames = []
+
+    bInfo = BenchmarkInfo(machineName=machineName, cudaVer="Test", osType="Test", pythonVer="Test", commitHash="Test")
+
+    for i in range(num):
+        db = ASVDb(asvDirName, repo, [branch1])
+        db.debugPrint = True
+        db.writeDelay=0.5
+        dbs.append(db)
+
+        funcName = f"somebenchmark{i}"
+        bResult = BenchmarkResult(funcName=funcName, result=43)
+        allFuncNames.append(funcName)
+
+        t = threading.Thread(target=db.addResult, args=(bInfo, bResult))
+        threads.append(t)
+
+    for i in range(num):
+        threads[i].start()
+
+    for i in range(num):
+        threads[i].join()
+
+    # There should be num unique results in the db after (re)reading.  Pick any
+    # of the db instances to read, they should all see the same results.
+    results = dbs[0].getResults()
+    assert len(results[0][1]) == num
+
+    # Simply check that all unique func names were read back in.
+    allFuncNamesCheck = [r.funcName for r in results[0][1]]
+    assert sorted(allFuncNames) == sorted(allFuncNamesCheck)
+
+    boto3.resource("s3").Bucket(bucketName).objects.filter(Prefix="asvdb/").delete()
```

Comment:
Not a huge deal but you might want to make a teardown function for this test that cleans up the bucket instead: https://docs.pytest.org/en/stable/xunit_setup.html
That way if one of the assertions above fails and the test exits early, the bucket still gets cleaned up.

Reply:
Ya, I ran into this while testing, will put this in.